server/src/main/java/org/elasticsearch/health/node/DiskHealthIndicatorService.java

```diff
@@ -0,0 +1,469 @@
+                List.of(),
+                List.of()
+            );
+        } else {
+            /*
+             * Something is unhealthy. We only report on the highest-priority problem. Basically:
+             *
+             * RED > YELLOW
+             * data nodes (including index locks from the cluster state) > master nodes > nodes of all other roles
+             *
+             * The highest priority problem is when we detect that there is an index with a block in the cluster state or a node that has
+             * a data role reporting a RED status (meaning it has exceeded the flood stage). This means that there is likely an index that
+             * cannot be written to.
+             * The next highest priority problem is when we detect that there is a non-data node that has a master role reporting a RED
+             * status (meaning it has exceeded the flood stage). This could lead to problems with cluster formation, discovery, etc.
+             * The next highest priority problem is when we detect that there is a non-data non-master node reporting a RED status
+             * (meaning it has exceeded the flood stage). This could lead to problems with certain features.
+             *
+             * The next tier down is when there are no nodes reporting RED, but some are reporting YELLOW. There will be no index
+             * blocks in the cluster state if we get here because the status would have automatically been RED. Similar to above, we only
+             * report problems with data nodes if there are any data nodes reporting YELLOW (the high watermark has been exceeded). If
+             * there are no data nodes reporting YELLOW, we report problems with any master nodes reporting YELLOW. If there are no
+             * master nodes reporting YELLOW, we report on the non-data non-master nodes that report YELLOW.
+             */
+            if (hasAtLeastOneIndexReadOnlyAllowDeleteBlock) {
+                /*
+                 * If there is an index block, we report RED, and report on any indices that are on RED nodes, or on any indices that are
+                 *  on YELLOW nodes if there are no RED nodes.
+                 */
+                final Set<String> nodesReportingRed = getNodesReportingStatus(diskHealthInfoMap, HealthStatus.RED);
+                Set<DiscoveryNodeRole> rolesOnRedNodes = getRolesOnNodes(nodesReportingRed, clusterState);
+                boolean atLeastOneRedDataNode = rolesOnRedNodes.stream().anyMatch(DiscoveryNodeRole::canContainData);
+                final Set<String> impactedNodes;
+                if (atLeastOneRedDataNode) {
+                    impactedNodes = getNodesWithDataRole(nodesReportingRed, clusterState);
+                } else {
+                    final Set<String> nodesReportingYellow = getNodesReportingStatus(diskHealthInfoMap, HealthStatus.YELLOW);
+                    impactedNodes = getNodesWithDataRole(nodesReportingYellow, clusterState);
+                }
```

Comment:
This seems to be discarding some nodes (RED non-data nodes for e.g). I think we should report all nodes that are problematic

Reply:
Yeah that was intentional based on a miscommunication we had. Now it reports on all of them.